### PR TITLE
drop `unused` destructuring declarations when `pure_getters` enabled

### LINKED ISSUE #105: Drop unused variables from destructuring assignments
**Bug report or Feature request?**
Not sure.

I'm trying to minize my build as mush as possible so I have `math.js` file in my project which exports all of the `Math` methods and props and some of my own stuff as well. Any time I need to do math I import funcs/vars from this file. I use rollup for bundling so there's no weird webpack module wiring or anything like that, all the code gets concatenated to something like:

```js
// from math.js
const { abs, acos, acosh /* ...etc */ } = Math
const lerp = () => /* ... */
const clamp = () => /* ... */

// from main.js
console.log(sin(PI))
```

And it goes to terser right after that. I assumed terser would figure out which variables the folowing code uses and which doesn't, so they can be dropped away safely. Turned out I was wrong, terser keeps all of them. I had some time trying to guess the right config options (like pure_funcs, pure_getters) but none of them seem to do the job. Especially on a code which has a destructuring assignment.

So I don't really know if this is the intended behavior or a bug. If a bug then it's a bug report. If not, then I'd like to know (and have) a way to drop everything that has no actual use in my code.

Here I have a few reproducible tests with unused vars. I don't really care about the one which uses aliasing instead of destructuring, but I tried it just in case and it also produced unexpected (for me) result.

**Version (complete output of `terser -V`)**

```
$ node -v
v10.9.0

$ npx terser -V
terser 3.8.2
```

**Complete CLI command or `minify()` options used**
```js
const terser = require('terser')

const options = {
  ecma: 8,
  compress: {
    // this one seems to convert `Math.foo` to `Math` (huh?)
    pure_getters: true,

    // this one seems to do nothing
    pure_funcs: Object
      .getOwnPropertyNames(Math)
      .filter(prop => typeof Math[prop] === 'function')
      .map(prop => 'Math.' + prop)
  }
}

const test = code => {
  const wrapped = `(() => { ${code} })()`
  const result = terser.minify(wrapped, options)
  console.log(result.code)
}

test(`
  const PI  = Math.PI
  const cos = Math.cos
  const sin = Math.sin
  const min = Math.min // ←
  const max = Math.max // ←

  console.log(cos(PI))
  console.log(sin(PI))
`)

test(`               // ↓↓↓  ↓↓↓
  const { PI, cos, sin, min, max } = Math

  console.log(cos(PI))
  console.log(sin(PI))
`)
```

**`terser` output or error**
```js
                                         // ↓↓↓↓ ↓↓↓↓
(()=>{const o=Math.PI,t=Math.cos,a=Math.sin;Math,Math;console.log(t(o)),console.log(a(o))})();
(()=>{const{PI:o,cos:n,sin:s,min:c,max:l}=Math;console.log(n(o)),console.log(s(o))})();
                          // ↑↑↑↑↑ ↑↑↑↑↑
```

**Expected result**
```js
// At least:
(()=>{const o=Math.PI,t=Math.cos,a=Math.sin;console.log(t(o)),console.log(a(o))})();
(()=>{const{PI:o,cos:n,sin:s}=Math;console.log(n(o)),console.log(s(o))})();

// Ideally: (convert to a destructuring assignment)
(()=>{const{PI:o,cos:n,sin:s}=Math;console.log(n(o)),console.log(s(o))})();
(()=>{const{PI:o,cos:n,sin:s}=Math;console.log(n(o)),console.log(s(o))})();
```

### PR DESCRIPTION
**Pull request by kzc**
fixes #105

### CODE REVIEW

**Comment by kzc** [comment]
@fabiosantoscode PTAL. This was more complicated than I first thought. See if I missed any corner cases.

**Comment by kzc** [comment]
This test case produces incorrect code with this PR:
```
$ echo 'export let [{x, y}] = [{x:1, y:2}];' | terser --toplevel -c pure_getters --ecma 6
export let[{}]=[{x:1,y:2}];
```
Need a better `export` check. I suppose I could do something like this:
```patch
--- a/lib/compress.js
+++ b/lib/compress.js
@@ -6649,7 +6649,7 @@ merge(Compressor.prototype, {
             && compressor.option("unused")
             && !self.is_array
             && Array.isArray(self.names)
-            && !has_ancestors(compressor, [/^VarDef$/, /^(Const|Let|Var)$/, /^Export$/])) {
+            && !compressor.find_parent(AST_Export)) {
             var keep = [];
             for (var i = 0; i < self.names.length; i++) {
                 var elem = self.names[i];
```
which would fix that test case but it would disable dropping unused variables within complex destructuring exports like the following:
```
$ echo 'export function f(){ let [{x, unused}] = [{x: 1, y: 2}]; return x;}' | terser --toplevel -bc pure_getters --ecma 6
export function f() {
    let [{x, unused}] = [ {
        x: 1,
        y: 2
    } ];
    return x;
}
```

**Comment by kzc** [comment]
I think this PR is good to go.

Other optimization opportunities noted in the tests can be handled in future PRs.

**Comment by fabiosantoscode** [comment]
LGTM. I think it's good to go too.
